### DISCUSSION TITLE
Command to calculate next version value

### DISCUSSION THREAD

**Discussion by laughedelic** [intuit/auto#1867] [Q&A] [answered]
Hi! I see that `auto version` return the semantic version bump, but is there a way to get the next (release) version with that bump applied? I think it would be very useful in non-standard release workflows that want to call atomic `auto` commands instead of relying on the complex ones like `shipit`.

**Comment by hipstersmoothie** [comment, marked as answer]
You can run `auto shipit --dry-run --quiet` and only the new version will be outputed.

**Reply by laughedelic** [reply]
This works, thanks! Would you be open to adding an option to the `version` command to make this more discoverable? e.g. `auto version --next` or something similar.

**Reply by hipstersmoothie** [reply]
From a code perspective i'm against adding it as a flag to the `version` command since it kinda tangles up the code too much for my liking.

This is a pretty common ask though.

I'm on the fence between:

- Add a new command (`calculate-version`) that just run `shipit -dq`
- Add more docs somewhere that alert people to the existence of `-dq` and it's use cases

I think a docs first approach might work better here. 

Would you be interested in contributing some docs around this area?

**Reply by laughedelic** [reply]
Thanks for response, I'll be willing to contribute, but more so to the implementation of the command/flag than the docs about a "workaround" (from my perspective anyway). Let me explain my point of view and please, forgive me if I'm misinterpreting something, I'm very new to this tool and I really want to understand it and be able to use it.

These lines on the docs front page resonate very well with me:

> `auto` makes no assumptions about your publishing process. Each tool is a function that can be run in isolation and _only does one thing really well_. For instance, you could just use `auto changelog` to generate the changelog and nothing else or use `auto version` to calculate just the `semver` bump.

But then `changelog` doesn't actually just generate a changelog but also writes it to a file and commits it (which is 3 different things), and `shipit` is a combination of several different multi-step release processes that depends on the context. 

I think all those things are useful, but when I read that paragraph I expected that there will be a set of simple commands that can be combined into any release process like building blocks. In that vein, when I saw the `version` command, I actually expected it to tell me _some version_ (current, last or next), not the _version bump_, just because it's called simply `version`, not `version-bump`. 

Why does the next version come from a command that encompasses the _whole release process_? This seems counter-intuitive to me. Doesn't the current `version` command have enough information for calculating the next version?

**Reply by hipstersmoothie** [reply]
I view most of the commands as they do 1 thing well. Commiting a changelog is a good part of making one, but I can see your point of view too. `shipit` definitely falls outside of that. That's why I refer to it as a `meta-command` in the docs, it will run all of the other parts of auto as needed.

> Doesn't the current version command have enough information for calculating the next version?

The version command is basically a `git log` from your last release to HEAD. It doesn't have any information about what the version _actually_ is. It's just looking at commits and trying to calculate the version bump. I see how the naming could be confusing. This name has been there for quite awhile and wasn't my decision. A better name would probably be `auto bump` but I feel we'd run into the same problem of people thinking that it would do the bump for them.

> Why does the next version come from a command that encompasses the whole release process?

`shipit` can make various types of releases (latest, next, canary, and old). It picks which one to create based on where it's run. `shipit` really just runs other commands under the hood (the commands for latest, next, canary, and old). So the utility in this is that it will give you the version for the appropriate release type.

You could you the `-dq` flags to just interact with those commands if you wanted to (ex: `auto canary -dq`) to better fit your release process.

**Reply by hipstersmoothie** [reply]
Really what it comes down to is that the publishing commands (shipit, latest, next, canary) are all heavily integrated with the package manager plugins. These plugins are where `auto` actually get your version from your project, but all interaction with that version is delegated to the plugins.

The "atomic" command (version, changelog, release) dont interact with the package manager plugins almost at all.

- `version` => Give the bump
- `changelog`=> Make a changelog given that bump
- `release` => Make a release for the last tag

As you can see none of this really requires interaction with the packages all that much. This is where the package manager plugins come in

- `version` 
- `changelog`
- Package manager plugin (via shipit) or user scripts facilitate a release
- `release`

**Reply by laughedelic** [reply]
Thanks! This is a great explanation!

I think I understand now where my confusion comes from. I thought that the last release/version information is normally taken from the git tags, but now I see that for someone coming from the Node.js background the source of truth for versioning is the package manager and its manifest file like `package.json`.

Okay, so we can document that to calculate a version one needs to run one of the release commands with the `--dry-run --quiet` flags. Some names are still quite confusing though in my opinion:
- `latest` gives the _next_ release version, while
- `next` gives the next _release candidate_ version

Again, I understand that they come from the npm conventions.

**Reply by bnigh** [reply]
Just came across this, and did not know this was a feature... so looks like these new discussions are working in at least 1 case  😄 

Regarding above, @hipstersmoothie , would you be able to clarify a few items:

(1)

> Why does the next version come from a command that encompasses the whole release process? This seems counter-intuitive to me. Doesn't the current `version` command have enough information for calculating the next version?

> `shipit` can make various types of releases (latest, next, canary, and old). It picks which one to create based on where it's run. `shipit` really just runs other commands under the hood (the commands for latest, next, canary, and old). So the utility in this is that it will give you the version for the appropriate release type.

I would agree that utilizing `shipit` initially seems a bit counter-intuitive. Could the `version incrementing` and `release type identification` logic be combined with `auto version` to correctly calculate the next version? or  does this just become the same as `shipit -dq` implementation?

A tangential thought:
- perhaps it would be useful to expose the `release type identification` logic as output of one of the cli commands (or a new one) since users may want to invoke additional logic for different release types outside of hooks (related to use case:  https://github.com/intuit/auto/discussions/1965)

(2) 

> I think I understand now where my confusion comes from. I thought that the last release/version information is normally taken from the git tags, but now I see that for someone coming from the Node.js background the source of truth for versioning is the package manager and its manifest file like `package.json`.

My current understanding is slightly different. Since `auto` initially reads from the tags / release notes, I was thinking the git tags / release notes were the source of truth (and writing to manifest file is side effect required by the package manager).

What does `auto` view as source of truth - the manifest versions, or the git tags / release notes?

**Reply by laughedelic** [reply]
@bnigh thanks for bringing this up again. I'm sort of trying to add an extra flag to the release commands as a shortcut to `--dry-run --quiet`: e.g. `auto shipit --version`. 

I'm curious in @hipstersmoothie's answers to your questions as well. Meanwhile, here are my two cents of understanding from reading the code:

> What does `auto` view as source of truth - the manifest versions, or the git tags / release notes?

I think it depends on the (publishing) plugins you enable. In case of `git-tag` plugin, git is the only source of information. But if you enable plugins like `npm`, `maven` or `gradle`, they read version value from the manifest (e.g. [here](https://github.com/intuit/auto/blob/main/plugins/npm/src/index.ts#L288)). There are two relevant hooks here:
- `getPreviousVersion` which does what it says
- `version` which calculates the _next_ (release) version using the previous version and the bump (based on the PRs and labels since the previous version) **and then** it also [creates the tag](https://github.com/intuit/auto/blob/main/plugins/git-tag/src/index.ts#L64-L70) and/or [modifies the manifest and commits it](https://github.com/intuit/auto/blob/main/plugins/npm/src/index.ts#L971-L978)

So the logic of calculating the next version and "setting" it is all in the same hook. It's possible to just get the version out of it if you run that hook with the `--dry-run` flag, but the hook is not available through the CLI (back to https://github.com/intuit/auto/discussions/1965#discussion-3338879).

To make things a bit more complicated: this `version` hook is used when you create a full release (`auto latest` or `shipit` on the main branch), but if you run `canary`, all the logic for calculating the `canary` version and publishing is in the `canary` hook, it doesn't use the `version` hook.

**Reply by bnigh** [reply]
@laughedelic 

> I'm sort of trying to add an extra flag to the release commands as a shortcut

A shortcut seems like should be simple and nice! In conjunction adding the details to docs should help, although admittedly, I'm not sure what would be most discoverable for docs. I'm not sure a user would immediately jump to reading flags to find next version. Like you, my initial thought was the `auto version` command. Perhaps is non-issue though if users can find via site-wide search.

---

> What does `auto` view as source of truth - the manifest versions, or the git tags / release notes?

Reading through more code, I'm a bit more confused...

I would have thought that auto would treat one of these as source of truth and solely read from that source, but it looks like it reads from both places:

- reads from tags:
  - during `auto version` and parent commands that invoke it (like `auto latest`), the semver bump is calculated from either the latest git tag or the latest github release when not using `from` flag
      - although these are technically different entities, I am treating as equivalent since it a github release requires a git tag
      - see: https://github.com/intuit/auto/blob/b1fcc83b780cbc581f8ead6a1b177af56a2c026f/packages/core/src/auto.ts#L1804...L1807
- reads from manifest
  - as you mentioned:
  > - `getPreviousVersion` which does what it says
  > - `version` which calculates the next (release) version using the previous version and the bump (based on the PRs and labels since the previous version) and then it also creates the tag and/or modifies the manifest and commits it
  
While ideally the manifest and tags should not be out of sync,

-  is it possible for a `version` hook to write and publish new version to manifest but fail to push tag (or vice-versa)?
    - If so, how would `auto` behave on next invocation?
    - Would semver bump starting ref (ie: `--from` field used in `auto version`) potentially be different than the ref of previous version found from `getPreviousVersion` hook?

**Comment by laughedelic** [comment]
For the reference, I just found this old issue with the same feature request: https://github.com/intuit/auto/issues/1490
